test/Constraints/keypath_swift_5.swift

```diff
@@ -19,3 +19,18 @@ func test() {
 
   let _ = C()[keyPath: \.i] // no warning for a read
 }
+
+
+struct T {
+    private(set) var a: Int
+    init(a: Int) {
+        self.a = a
+    }
+}
+
+func testReadOnlyKeyPathDiagnostics() {
+  let path = \T.a
+  var t = T(a: 3)
+  t[keyPath: path] = 4 // expected-error {{cannot assign through subscript: 'path' is a read-only key path}}
+  t[keyPath: \T.a] = 4 // expected-error {{cannot assign through subscript: the key path is a read-only key path}}
```

Comment:
`the key path is a read-only key path` reads a bit weird to me - maybe this could be `'T.a' is a read-only key-path` and then it would match the diagnostic above.

Reply:
The current behavior for similar errors is to give the error `"cannot assign through subscript: 'x' is a read-only key path"` if there's a declaration that we can get the name from, and give "cannot assign through subscript: immutable key path". As such, adding additional information to the error message in the second case would be an improvement aside from the bug fix. I'd be happy to change the wording as I agree it is a bit clunky (maybe just "cannot assign through subscript: is read-only" might be better?) but I feel like adding the keyPath name here might be outside the scope of this PR. Thoughts @xedin?

Reply:
Yeah, I think it would make sense to just drop "key path" at the end, so it would say `key path is read-only`.